src/v/storage/log_reader.cc

```diff
@@ -268,6 +269,17 @@ log_reader::do_load_slice(model::timeout_clock::time_point timeout) {
         return _iterator.close().then(
           [] { return ss::make_ready_future<storage_t>(); });
     }
+    /**
+     * We do not want to close the reader if we stopped because requested range
+     * was read. This way we make it possible to reset configuration and reuse
+     * underlying file input stream.
+     */
+    if (
+      _config.start_offset > _config.max_offset
+      || _config.bytes_consumed > _config.max_bytes || _config.over_budget) {
+        set_end_of_stream();
+        return ss::make_ready_future<storage_t>();
```

Comment:
the destructor should have a vassert() that it was closed

Reply:
it is already there in `log_reader` destructor